### ISSUE TITLE
prerelease strategy should take git-flow into account

### ISSUE THREAD

**Issue by link89** [semantic-release/semantic-release#1231]
## New feature motivation
Many project using git-flow since it is more friendly to large development team.
https://datasift.github.io/gitflow/IntroducingGitFlow.html

But for now semantic-release@16.0.0-beta.22 seems didn't handle this workflow properly.
The major reason is semantic-release always read tag from current branch.
In gitflow, feature branch/release branch is always folked from develop branch instead of from master, and develop branch never rebase to master. 
Which make it impossible to read tag from master.

## New feature description

In order to solve this problem. I suggest we should read tags of release branches when deciding latest release.

**Comment by link89** [comment]
Hi  @pvdlg 
Can you should me some light about how to using semantic-release with gitflow?
Thanks in advance.

**Comment by MarcLoupias** [comment]
The widely used GitFlow is not a git workflow CI/CD friendly. It is mostly obsolete.

The purpose of the `release/x.y.z` branches is to add a QA step, mostly manual in practice which is in complete opposition with the Continuous Delivery where the automation is the key. If you don't trust your automated testing suite to deploy you cannot do Continuous Delivery.

Furthermore, one of the purpose of semantic-release is to let the tool generate the version number based on the commit's messages.

In GitFlow you determine the version number before creating the `release/x.y.z` branch.

This is a major incompatibility.

Another incompatibility point is that GitFlow use several integration branches, and CI/CD workflows needs to have just a single one. An integration (or collaboration) branch is a branch used as a target for Merge Request / Pull Request. In GitFlow you have `release/x.y.z` branches, `hotfix/x.y.z` branches and `develop` to build your next `release/x.y.z` branch. This problem is described as the [Concurrent development of consecutive releases](https://trunkbaseddevelopment.com/concurrent-development-of-consecutive-releases/). `hotfix/x.y.z` branches are releases (you will deploy them), and `release/x.y.z` branches are feature frozen scope of the `develop` branch.

Semantic-release is designed to do Continuous Deployment by default (**Each** commit triggers a deploy), and you can do Continuous Delivery (Each commit **can** be deployed when you want to) by adding a `develop` branch where you target your MR/PR and you leave your `master` branch as a vector only for deployment. In that case your `master` branch is not a collaboration branch anymore, just a vector to trigger a deploy.

If you want to speed up your [lead time](https://devops.com/metrics-devops/) you should upgrade to a more modern git workflow like [trunk based development](https://trunkbaseddevelopment.com/) or [GitHub Flow](https://guides.github.com/introduction/flow/). Both are compatibles with semantic-release.

**Comment by pvdlg** [comment]
Thanks @MarcLoupias for the detailed explanation.

semantic-release is indeed not compatible with Git flow and primarily meant to be used with the [GitHub Flow](https://guides.github.com/introduction/flow/). As a side note, in v15.0.0 or v16.0.0 is you don't use multiple release branches there was workarounds to make semantic-release partially works with Git Flow but that was never the intended way of using it.

semantic-release is a tool designed to automate continuous delivery, as such:
- it determines the release number automatically
- it assumes that each commit pushed or merged to a release branch is ready to be release and therefore expect you to have a robust test suite to ensure that
- release each commit pushed or merged to a release branch right away

Git Flow is the exact opposite, it let you manage the content and the timing of your release manually via release branches. It's a very different way of working that as it own set of advantages and inconvenient, just like any other.

As part of the workflow management, semantic-release also provides automation tools that allow to  creates Git tags, publish to npm, generate changelog, create GitHub releases etc... I suspect you are interested in those pieces of semantic-release and try to integrate them into a Git Flow.
If it's actually the case I would recommend you to not use semantic-release and instead use more specific tools for those automation steps that you could integrate in your Git Flow. Here is a few example of tools you could investigate: [zeit/release](https://github.com/zeit/release), [release-it](https://github.com/release-it/release-it), [github-tools/github-release-notes](https://github.com/github-tools/github-release-notes), [conventional-changelog](https://github.com/conventional-changelog/conventional-changelog), [np](https://github.com/sindresorhus/np).

**Comment by nolde** [comment]
> In GitFlow you determine the version number before creating the release/x.y.z branch.

Not necessarily. I have been using *gitflow* + `semantic-release` for a while now with great success. Gitflow just determines a branch workflow, and that can be adapted to several different needs.

I have a full CI/CD pipeline that works using gitflow, and allows us to have stable QA environment for our testing team, and versioning is fully controlled by `semantic-release`. Our release branches are just called `release`, or `release/next` or `release/hotfix`, and they deploy to testing environments. Once testing is finalised, they the release branch is merged into master, generating the final version and deploying to production.

The only problem we had was with _ prerelease_ versioning, which we solved by appending CI/CD pipeline execution and commit hash. Now, I am actually moving towards using [`@saithodev/semantic-release-backmerge`](https://github.com/saitho/semantic-release-backmerge) plugin, which allows for easy and automated sync of tags between master and develop.

I am not arguing for adding any features to `semantic-release`, I just want to leave this here for anyone trying to use *gitflow*  with `semantic-release`.

**Comment by MarcLoupias** [comment]
@nolde This is because you are developing only one release at once which is really comfortable but also very rare.

Most of the teams will develop several versions in parallel which means several develop branches, several releases/hotfix branches and they will fall in the merging hell.

That's why organizations doing true agile comes to trunk-based development aka continuous integration. And for that need GitFlow is irrelevant.

**Comment by nolde** [comment]
@MarcLoupias There is nothing obsolete or irrelevant about gitflow. I know that's a common subject nowadays, but there are just different tools for different needs and teams.

I have a fleet of micro-services in a almost complete serverless environment, and gitflow serves my need very well; it allows me to develop different microservice that might depend on each other (depending on scale of change) and merge then into development branch, instead of forgotten left-over feature branches. Also, allows for creation of the most basic development environment for our team.

When we're ready for testing, all code gets pulled into a release branch, which deploys code to a more stable environment, allowing QA team not to be affected by random developer changes. When all is deemed ok, code is merged into master and automatically deployed to production.

Again, this is already closed, so I won't chime in again, but [backmerge](https://github.com/saitho/semantic-release-backmerge) solves most of my issues with semantic-release and gitflow, and I am therefore quite happy.

**Comment by stoopman** [comment]
> @MarcLoupias There is nothing obsolete or irrelevant about gitflow. I know that's a common subject nowadays, but there are just different tools for different needs and teams.
> 
> I have a fleet of micro-services in a almost complete serverless environment, and gitflow serves my need very well; it allows me to develop different microservice that might depend on each other (depending on scale of change) and merge then into development branch, instead of forgotten left-over feature branches. Also, allows for creation of the most basic development environment for our team.
> 
> When we're ready for testing, all code gets pulled into a release branch, which deploys code to a more stable environment, allowing QA team not to be affected by random developer changes. When all is deemed ok, code is merged into master and automatically deployed to production.
> 
> Again, this is already closed, so I won't chime in again, but [backmerge](https://github.com/saitho/semantic-release-backmerge) solves most of my issues with semantic-release and gitflow, and I am therefore quite happy.

Hi @nolde could you please share your semantic-release config and ci/cd setup which made it work with gitflow? Much appreciated!

**Comment by regicsolutions** [comment]
@nolde I am also making use of GitFlow and interested in leveraging semantic-release to automate our release workflow, would it be possible to share a sample of your semantic-release configuration?

**Comment by nolde** [comment]
Sure. 

The code below is a simplified version of my current setup. I use bitbucket, and bitbucket pipelines to deploy our code.

Details:
- I have slightly different versions depending on type of project (web applications, backend services, node.js libraries). This one is for web applications.
- The shell script is used basically to upload the final bundle to bitbucket's "Downloads" section, kind of mimicking Github's Releases. It also deploys the deliverables to the right place.
- QA environment allows so `chore`s actually trigger a semantic-release build and version update. 
- The application build is done inside semantic-release process so I can carry the proper application version.
- PROD builds do a backmerge into `develop`, keeping all tags sync'ed.

```javascript
const CREDENTIALS = process.env.GIT_CREDENTIALS
const OWNER = process.env.BITBUCKET_REPO_OWNER
const REPO = process.env.BITBUCKET_REPO_SLUG
const STAGE = process.env.STAGE

// //////////////////////////////////////////////

const changelogFile = 'CHANGELOG.md'
const writerOpts = { commitsSort: ['scope', 'subject'] }
const prepareCmd = `npm run build:${STAGE}`
const release = 'sh .release.sh ${nextRelease.version}'
const preset = 'conventionalcommits'
const qaReleaseRules = [{ type: 'chore', release: 'patch' }]
const presetConfig = {
  compareUrlFormat:
    '{{host}}/{{owner}}/{{repository}}/branches/compare/{{currentTag}}..{{previousTag}}'
}

// //////////////////////////////////////////////

module.exports = {
  repositoryUrl: `https://${CREDENTIALS}@bitbucket.org/${OWNER}/${REPO}`,
  branches: [
    {
      name: 'master'
    },
    {
      name: 'release/*',
      prerelease: 'rc',
      channel: 'next'
    },
    {
      name: 'hotfix/*',
      prerelease: 'fix',
      channel: 'hotfix'
    }
  ],
  plugins:
    STAGE === 'qa'
      ? [
          // QA (order matters)
          ['@semantic-release/commit-analyzer', { preset, releaseRules: qaReleaseRules }],
          ['@semantic-release/npm', { npmPublish: false }],
          ['@semantic-release/release-notes-generator', { preset, presetConfig, writerOpts }],
          ['@semantic-release/exec', { prepareCmd }],
          ['@semantic-release/git']
        ]
      : [
          // PROD (order matters)
          ['@semantic-release/commit-analyzer', { preset }],
          ['@semantic-release/npm', { npmPublish: false }],
          ['@semantic-release/release-notes-generator', { preset, presetConfig, writerOpts }],
          ['@semantic-release/changelog', { changelogFile }],
          ['@semantic-release/exec', { prepareCmd, successCmd: release }],
          ['@semantic-release/git'],
          [
            '@saithodev/semantic-release-backmerge',
            {
              clearWorkspace: true,
              backmergeStrategy: 'merge',
              mergeMode: 'ours'
            }
          ]
        ]
}
```

**Comment by regicsolutions** [comment]
This is super helpful @nolde thank you, cheers!

**Comment by regicsolutions** [comment]
@nolde would it be possible to get a sample of your release.sh script? am also using Bitbucket.

**Comment by nolde** [comment]
Sure.

This script upload a bundle, and then triggers another pipeline to make the actual deployment out of the uploaded bundle. So you may only need half of it.

```bash
set -e

export BITBUCKET_COMMIT=$(git rev-parse HEAD | awk '{print substr($0,1,7);exit}')
NEW_VERSION=v\${nextRelease.version}.$BITBUCKET_BUILD_NUMBER.$BITBUCKET_COMMIT

tar -cvzf "$NEW_VERSION.tar.gz" <files and folders you want as part of the bundle>
echo "Artifact bundle created"

curl \
  -X POST -s -u $GIT_CREDENTIALS \
  https://api.bitbucket.org/2.0/repositories/$BITBUCKET_REPO_OWNER/$BITBUCKET_REPO_SLUG/downloads \
  -F files=@"$NEW_VERSION.tar.gz"
echo "Artifact bundle uploaded"

curl \
  -X POST -is -u $GIT_CREDENTIALS \
  -H 'Content-Type: application/json' \
  https://api.bitbucket.org/2.0/repositories/$BITBUCKET_REPO_OWNER/$BITBUCKET_REPO_SLUG/pipelines/ \
  -d '{
    "target": {
      "selector": {
        "type": "custom",
        "pattern": "production-deploy"
      },
      "ref_name": "master",
      "ref_type": "branch",
      "type": "pipeline_ref_target"
    },
    "variables": [{
      "key": "VERSION",
      "value": "'"$NEW_VERSION"'",
      "secured": false
    }]
  }'
echo "Deployment triggered"
```

**Comment by regicsolutions** [comment]
Awesome, thanks @nolde cheers!

**Comment by regicsolutions** [comment]
@nolde started playing around with the configuration you shared today and had a couple of questions if you don't mind:

1. Are you manually creating release/v1.x.x branches from the develop branch and triggering a release by initiating that new branch and committing a `chore` or `patch` message? Was hoping to automate the creation of the release branch using the same message while I am in my develop branch. 

2. Are you creating a git tag using the latest version after you merge release/v1.x.x to master? I see the back-merge plugin is doing the merge back from master but was unsure where develop is getting defined and where in the flow is the git tag being created?

**Comment by nolde** [comment]
@regicsolutions No problem at all.

1. I just create a `release/next` or `release/thisMoreSpecificRelease` or anything I want, really. If you want, you could event remove the level requirement. The version creation is fully defined by `semantic-release`, using the normal rules; I've just added `chore` to it to allow small fixes to go into QA release candidate without creating an extra changelog line in the upcoming production release. You can remove that if you don't need or agree with it.

2. Tags are fully taken care of by `semantic-release` as well; there's no manual tagging of any kind. The `develop` branch is defined within the backmerge plugin itself, as it is the default development branch name in gitflow. You can modify it to anything you want, [check the plugin's docs](https://github.com/saitho/semantic-release-backmerge). Tags are created following the `semantic-release` flow, just check the docs.

**Comment by GuilhermeGomes1929** [comment]
@nolde How do you increment the numbers in `semantic-release` using `Gitflow`?
Do you increment the `minor` number every single `feature` or every `release`?
Because the `minor` is used when you add some new functionality that is compatible. And the `release branch` is a group of new functionalities, correct?

**Comment by nolde** [comment]
@GuilhermeGomes1929 

I am not sure I understand your question, but I use `gitflow` as a concept; I don't use any specific tools with it other than `git` itself. I don't determine versions at all, as that is determined by `semantic-release` -- which is the point of using it -- by reading my commits. 

When I want to release, I just branch off `develop` with the appropriate branch name convention for a release branch in my setup, and `semantic-release` takes care of the rest.

**Comment by webmutation** [comment]
@nolde Since you mention you are doing Java microservices, do you use SNAPSHOT versions? 

I am currently struggling with supporting the publishing and tagging of SNAPSHOT versions, it seems it is only possible to keep track of one tag at a time, or for prerelease but those are not tagged appropriately for JAVA Snapshot publishing. If you have experienced a similar issue, perhaps you already have a way to solve it. It seems I will have to add helper functions to be able to use semantic-release with SNAPSHOT

**Comment by nolde** [comment]
@webmutation Unfortunately, my projects are JavaScript based (SPAs for frontends, nodejs microservices for backend), not Java and thus, cannot really help you with that one.

That said, the configuration I've pasted above should potentially support any kind of variation between staging, prerelease or production with some fine-tuning.

**Comment by webmutation** [comment]
@nolde Thanks for your feedback... I must have misread you were using Java. 

I got stuck with the existing plugins, they do not support SNAPSHOT releases. I guess I will have to do it via scripting 

This is my latest attempt to try and make it work without having to write too much additional code.

```
{
  "branches": ["main", {name: "develop", channel: "develop", prerelease: "SNAPSHOT"}],
  "plugins": [
    "@semantic-release/commit-analyzer",
    "@semantic-release/release-notes-generator",   
    "@semantic-release/changelog",
    [
      "@semantic-release/exec",
      { 
        "verifyReleaseCmd": 'mvn versions:set -DnewVersion="${nextRelease.version}"'
      },
    ],    
    "@semantic-release/gitlab",
    [
      "@terrestris/maven-semantic-release",
      {
        "clean": false,
        "updateSnapshotVersion": true,
        "snapshotCommitMessage": "chore(release): Updating SNAPSHOT [skip ci]",
        "settingsPath": "/builds/apim/cc-tester-backend/.m2/settings.xml",
      }
    ],
    "@semantic-release/git",
      {
      "assets": ["CHANGELOG.md","pom.xml"],
      "message": "chore(release): ${nextRelease.version} [skip ci]",
      }
  ]
}
```

I have two remaining problems, the maven-semantic-release does not have a releaseSnapshotVersion method.
Also for some reason, the pom.xml file is not committed by the git plugin... 

I will try to adapt the code you posted above, I will have to write the snapshot release anyway... also the updateSnapshotVersion breaks with the branching naming...
